dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactory.java

```diff
@@ -0,0 +1,129 @@
+package com.google.edwmigration.dumper.application.dumper.connector.teradata;
+
+import static com.google.edwmigration.dumper.application.dumper.connector.teradata.AbstractTeradataConnector.SQL_FORMAT;
+
+import com.google.common.base.Joiner;
+import com.google.edwmigration.dumper.application.dumper.connector.SqlQueryFactory;
+import com.google.edwmigration.dumper.application.dumper.connector.ZonedInterval;
+import java.util.List;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+
+public class TeradataSqlQueryFactory implements SqlQueryFactory {
+  private final ZonedInterval interval;
+  private final String logTable;
+  private final String queryTable;
+  private final String errorCheckStatement;
+  private final String baseTableName;
+  private final String timeFieldName;
+  private final List<String> conditions;
+  private final List<String> orderColumns;
+  private final List<String> expressions;
+  private final boolean joinQueryTable;
+
+  /* pp */ TeradataSqlQueryFactory(
+      ZonedInterval interval,
+      String logTable,
+      String queryTable,
+      String errorCheckStatement,
+      String baseTableName,
+      String timeFieldName,
+      List<String> conditions,
+      List<String> orderColumns,
+      List<String> expressions,
+      boolean joinQueryTable) {
+    this.interval = interval;
+    this.logTable = logTable;
+    this.queryTable = queryTable;
+    this.errorCheckStatement = errorCheckStatement;
+    this.baseTableName = baseTableName;
+    this.timeFieldName = timeFieldName;
+    this.conditions = conditions;
+    this.orderColumns = orderColumns;
+    this.expressions = expressions;
+    this.joinQueryTable = joinQueryTable;
+  }
+
+  private String getTableFrom(String columnName) {
+    return isQueryTable(columnName) ? queryTable + " ST" : logTable + " L";
+  }
+
+  /**
+   * For each potential expression in EXPRESSIONS, work out whether this Teradata accepts it, and if
+   * so, use it as part of the eventual query.
+   *
+   * @param predicate A predicate to compute whether a given expression is legal.
+   * @return A SQL query containing every legal expression from EXPRESSIONS.
+   */
+  @Override
+  public String getSql(BiPredicate<String, String> predicate) {
```

Comment:
The list of available columns doesn't change during a single dumper invocation. The predicate is passed to the query building logic in the existing code, but it is a bit of a hack. I would suggest to decouple filtering of available columns from query building, so that this query factory can accept a list of available columns.